tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -1222,3 +1115,170 @@ async def _compute_intrinsic_federated_collect(self, arg):
   @tracing.trace
   async def _compute_intrinsic_federated_secure_sum(self, arg):
     return await self.intrinsic_strategy.federated_secure_sum(arg)
+
+
+class Channel:
+
+  def __init__(self, sender_placement, receiver_placement, is_encrypted):
+
+    self.sender_placement = sender_placement
+    self.receiver_placement = receiver_placement
+    self.is_encrypted = is_encrypted
+
+  async def setup(self, parent_executor):
+
+    if not self.is_encrypted:
+      return
+
+    pk_receiver, sk_receiver = await self._receiver_generate_keys(
+        parent_executor)
+
+    return pk_receiver, sk_receiver
+
+  async def send(self, parent_executor, val, pk_receiver):
+
+    if not self.is_encrypted:
+      return val
+
+    return await self._encrypt_sender_tensors(parent_executor, val, pk_receiver)
+
+  async def receive(self, parent_executor, val, sk_receiver, sender_dtype):
+
+    if not self.is_encrypted:
+      return val
+
+    return await self._decrypt_tensors_on_receiver(parent_executor, val,
+                                                   sk_receiver, sender_dtype)
+
+  async def _receiver_generate_keys(self, parent_executor):
+    # NOTE solve confusion between executor and receiver_executor
+
+    @computations.tf_computation()
+    def generate_keys():
+      pk, sk = easy_box.gen_keypair()
+      return pk.raw, sk.raw
+
+    fn_type = generate_keys.type_signature
+    fn = generate_keys._computation_proto
+
+    receiver_ex = parent_executor._get_child_executors(
+        self.receiver_placement, index=0)
+
+    key_generator = await receiver_ex.create_call(await
+                                                  receiver_ex.create_value(
+                                                      fn, fn_type))
+
+    keys = await asyncio.gather(*[
+        receiver_ex.create_selection(key_generator, i)
+        for i in range(len(key_generator.type_signature))
+    ])
+
+    pk_fed_val, sk_fed_val = await asyncio.gather(
+        *[parent_executor._place(k, self.receiver_placement) for k in keys])
+
+    # NOTE: This broadcast should work from aggregator to clients
```

Comment:
this will not work until aggregators is a proper placement, since the compiler needs to recognize it as such in order to find the right IntrinsicDef to compute. I recommend we switch to the more explicit `_place`

Reply:
agreed - makes sense.